src/serializer.ts

```diff
@@ -127,35 +127,54 @@ export class Serializer {
      * This will print a JSON string for `obj`, without `password` property in it, and `foo` property renamed to `bar`.
      *
      * @param data
+     * @param additionalData additional data you want to add to keep trace of a context during serialization in a
+     * child class, the goal behind that is to provide a way for serializers extending this one to use custom data
+     * across a single object, no matter how deep we are in the object.
      * @returns {string}
      */
-    public serialize(data: any): string {
+    public serialize(data: any, additionalData?: any): string {
         //We have to create a clone of data in order to avoid deleting properties in the original object.
-        const obj = JSON.parse(JSON.stringify(data));
+        const obj = this.prepareSerialize(JSON.parse(JSON.stringify(data)), data, additionalData);
+        return JSON.stringify(obj);
+    }
+
+    protected prepareSerialize(obj: any, instance: any, propertyKey?: string, additionalData?: any): any {
```

Comment:
The signature of this method bothers me. It handles two different case, but with one method. You declared 4 params but only used 3 in the first call, and the param name does not match. I didn't go too deep in the implementation, but in this case, I think you should declare the different signature, and then the implementation.

_This is an example_

```ts
protected prepareSerialize(obj: any, instance: any, additionalData: any): any;
protected prepareSerialize(obj: any, instance: any, propertyKey?: string): any;
protected prepareSerialize(obj: any, instance: any, propertyKey?: string, additionalData?: any): any {
   // The real implementation
}
```

Reply:
The problem with these custom signatures is the fact that the recursive call (L151) isn't valid as it doesn't match any of the signatures (expected 2-3 arguments, but got 4).